clusterloader/docs/vision.md

```diff
@@ -0,0 +1,344 @@
+
+### Config
+
+A single test scenario will be defined by a `Config`. Its schema will be as following:
+
+```
+struct Config {
+	// TBD: Some config allowing to access cluster, e.g. path to kubeconfig file.
+
+	// Number of namespace in the test.
+	Namespaces int32
+	// Steps of the test.
+	Steps []Step
+}
+```
+
+With a test being defined by a single json/yaml file, it should be pretty simple
+to modify scenarios, and fork them to new ones.
+
+Note that before running any steps, ClusterLoader will create all the necessary
+namespaces and after running all of them will delete them (together with all
+objects that remained undeleted after running the test). Namespaces are described
+in more details in the later part of this document.
+
+### Step
+
+Each step will consist of a number of create, update and delete operations (potentially
+on many different object types) or a number of monitoring/measurement-related actions.
+A single step is defined as following:
+
+```
+struct Step {
+	// Only one of these can be non-empty.
+	Phases       []Phase
+	Measurements []string
+}
+```
+
+We make `Phases` and `Measurements` separate concepts to ensure correct ordering
+between those two types of actions. It's very important to ensure that proper
+measurements are started before we start given actions and they are stopped when
+all actions are done.
+
+Also note that all `Phases` and `Measurements` within a single `Step` will be
+run in parallel - a `Step` ends when all its `Phases` or `Measurements` finish.
+
+### Phase
+
+Phase declaratively defines state of objects we should reach in the underlying
+Kubernetes cluster. A single declaration may result in a number of create, update
+and delete operations depending on the current state of the cluster.
+
+We define the phase as following:
+
+```
+struct Phase {
+	// Set of namespaces in which objects should be reconciled.
+	// If null, objects are assumed to be cluster scoped.
+	NamespaceRange *NamespaceRange
+	// Number of instances of a given object to exist in each
+	// of referenced namespaces.
+	ReplicasPerNamespace int32
+	// Name of TuningSet to be used.
+	TuningSet string
+	// A set of objects that should be reconciled.
+	Objects []Object
+}
+```
+
+```
+struct Object {
+	// Type definition for a given object.
+	ObjectType ObjectType
+	// Base name from which names of objects will be created.
+	// Names of objects will be "basename-0", "basename-1", ...
+	Basename string
+	// A file path to object definition.
+	ObjectTemplatePath string
+}
+```
+
+The semantic of the above structure will be as following:
+- `Phases` within a single `Step` will be run in parallel
+- `Object` within a single `Phase` will be reconciled in serial for a given
+  (namespace, replica number pair). For different (namespace, replica number)
+  pairs they will be spread using a given tuning set.
+
+The rationale for having such structure is the following:
+- Slice of `Object` allows for making a unit of work to be greater than
```

Comment:
I'd suggest rewording this sentence as:

`Objects` represents a collection of k8s objects that can be logically thought of as a unit of workload (for e.g a web application that's comprised of a service, a deployment and a volume). Conceptually, this collection is our unit for replication. The replication itself is done according to the TuningSet and ReplicasPerNamespace parameters of the `Phase`. Also note that we process the `Objects` slice serially which allows ordering while processing objects of a unit (e.g create service before creating deployment).

Reply:
Slightly modified - but thanks for making it clear.